python/lsst/ts/linearstage/config_schema.py

```diff
@@ -12,80 +11,93 @@
         description: >
           Maximum target position in mm
         type: number
-        default: 70
     serial_port:
         description: USB port for the serial interface
         type: string
-        default: "/dev/ttyLinearStage1"
     daisy_chain_address:
         description: The daisy-chain device address as located in the daisy chain (for Zaber Devices only)
         type: number
-        default: 1
     steps_per_mm:
         description: This is approximately the amount of steps in a millimeter (for the Zaber stage only).
         type: number
-        default: 8000
     socket_address:
         description: The IP address to establish a socket connection (for use with Igus Dryve v1 controllers)
         type: string
-        default: "0.0.0.0"
+        format: hostname
     socket_port:
         description: >
           The network port to establish a socket connection (for use with Igus Dryve v1
           controllers)
         type: number
-        default: 502
     feed_rate:
         description: >
           Distance of travel [mm] per single motor rotation (for use with Igus Dryve v1
           controllers)
         type: number
-        default: 150
     maximum_stroke:
         description: Maximum travel distance (stroke) from the homed position [mm]
         type: number
-        default: 1000
     homing_speed:
         description: >
           Speed to use for homing in millimeters per second [mm/s] (for use with Igus Dryve v1
           controllers)
         type: number
-        default: 20.0
     homing_acceleration:
         description: >
           Acceleration to use for homing in millimeters per second squared [mm/s^2] (for use with
           Igus Dryve v1 controllers)
         type: number
-        default: 5
     homing_timeout:
         description: >
           Amount of time to wait for homing to complete before timing out.
         type: number
-        default: 60
     motion_speed:
         description: >
           Speed to use for standard travel motion in millimeters per second [mm/s] (for use with
           Igus Dryve v1 controllers)
         type: number
-        default: 30.0
     motion_acceleration:
         description: >
           Acceleration to use for standard travel motion in millimeters per second squared [mm/s^2]
           (for use with Igus Dryve v1 controllers)
         type: number
-        default: 5
+allOf:
+  - if:
+      properties:
+        stage_type:
+          const: "Igus"
+    then:
+      required:
+        - socket_address
+        - socket_port
+        - feed_rate
+        - maximum_stroke
+        - homing_speed
+        - homing_acceleration
+        - motion_speed
+        - motion_acceleration
+  - if:
+      properties:
+        stage_type:
+          const: "Zaber"
+    then:
+      required:
+        - target_position_minimum
+        - target_position_maximum
+        - serial_port
+        - daisy_chain_address
+        - steps_per_mm
```

Comment:
This does not work well with override files. The issue is that `_init.yaml` must decide which device it is providing configuration for, and an override may want to pick the other device and has to start over.

Please file a Jira ticket to improve this (or fix it on this ticket, if you prefer). The two common solutions are:

1: If you know you will only ever have 2 or 3 devices, then include full configuration for all devices, e.g. add zaber_config and igus_config sections with the appropriate properties for each. That way `_init.yaml` can provide full configuration for both, and an override file can tweak only the parameters it cares about. This is optimal for discovery in `_init.yaml` and is probably what you want here.

2: If you expect to have more kinds of linear stages in the future then consider using pluggable algorithms. One class handles Zaber, another handles Igus, and each has its own config schema. Look at ts_watcher rules and ts_ess_csc data clients for examples. This scales well to additional devices, and is handy if most device properties have sensible default values. One big advantage is that algorithm-specific configs *can* have default values; in best cases this means that the amount of device-specific config the user has to provide is minimal.

As a reminder (sorry if this is too obvious), in all cases:
* You must increment the version (in the  `title`).
* You still need the `devices` property, or something like it, to specify which device you are actually using.
* All properties that will be common to all devices should still be at the top level, rather than buried in the device-specific sections.

Reply:
Filed ticket to improve configuration schema.